text/0000-macro-hygiene-optout.md

```diff
@@ -0,0 +1,229 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Escaping of hygiene for identifiers within macros allows one to define identifiers with syntax contexts (**hygiene**) corresponding to the location in the source code from which the macro is invoked (the **call-site**) rather than the location it is defined (**definition-site**). It also enables one to use/reference existing identifiers from the call-site from within macro definitions, though this is not the true aim of the feature, but rather a side-effect, and will be discussed later.
+
+Note that for the purposes of this RFC, an **identifier** can roughly be considered to be an textual name (e.g. `foo_bar`) of any sort (for a variable, function, trait, etc.) or a lifetime (e.g. `'a`).
+
+To escape an identifier in code, one simply prefixes an identifier with the [sigil] `#`. This changes the syntax context (hygiene) of the identifier from the usual definition-site to the call-site.
+
+## Guide: Example A
+[guide-example-a]: #guide-example-a
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m() {
+    pub mod #foo {
+        pub const #BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!(); // `foo` and `foo::BAR` both behave as if they were defined directly here.
+    assert_eq!(123, foo::BAR);
+}
+```
+
+## Guide: Example B
+[guide-example-b]: #guide-example-b
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m($mod_name:ident) {
+    pub mod $mod_name {
+        pub const #BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!(foo); // `foo` and `foo::BAR` both behave as if they were defined directly here.
+    assert_eq!(123, foo::BAR);
+}
+```
+
+## Guide: Example C
+[guide-example-c]: #guide-example-c
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m($mod_name:ident) {
+    pub mod $mod_name {
+        pub const BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!(foo);
+    let _ = foo::BAR;
+    //~^ ERROR cannot find value `BAR` in module `foo`
+}
+```
+
+## Guide: Example D
+[guide-example-d]: #guide-example-d
+
+```rust
+#![feature(decl_macro)]
+#![feature(macro_hygiene_optout)]
+
+macro m() {
+    pub mod #foo {
+        pub const BAR: u32 = 123;
+    }
+}
+
+fn main() {
+    m!();
+    let _ = foo::BAR;
+    //~^ ERROR cannot find value `BAR` in module `foo`
+}
+```
+
+## Meta-variables
+[meta-variables]: #meta-variables
+
+Hygiene escaping of meta-variables (i.e. `#$foo` and `$#foo`) does not have immediately obvious semantics or usefulness, so is explicitly disallowed for the present, and yields error messages.
```

Comment:
The obvious semantics to me is that the resulting identifier takes the name from the metavariable and the hygiene context from the call site.

Reply:
Yes, I really meant the former in't obviously useful, why the latter isn't obviously useful either *nor* does it have obvious semantics.